pkg/cloudevents/types/value.go

```diff
@@ -0,0 +1,210 @@
+// x can be any type that is convertible to a CloudEvents attribute.
+func StringOf(x interface{}) (string, error) {
+	v, err := ValueOf(x)
+	return v.String(), err
+}
+
+// Value holds the value of a CloudEvents attribute.
+//
+// Value can hold the native type or the string representation.
+// The To...() methods extract the value as a native type.
+type Value struct{ v interface{} }
+
+// Interface returns the value as an interface{}.
+//
+// The return value may be nil if the Value is empty, otherwise it will always
+// be one of the native types: bool, int32, string, []byte, url.URL, time.Time.
+func (v Value) Interface() interface{} { return v.v }
+
+// ValueOf validates the type and range of x and wraps it in a Value.
+// x can be any type that is convertible to a CloudEvents attribute.
+func ValueOf(x interface{}) (Value, error) {
+	switch x := x.(type) {
+
+	case Value: // Already a Value, no wrapping
+		return x, nil
+
+	case bool, int32, string, []byte, url.URL, time.Time:
+		return Value{x}, nil // Already a preferred type, no conversion needed.
+
+	case uint, uintptr, uint8, uint16, uint32, uint64:
+		u := reflect.ValueOf(x).Uint()
+		if u > math.MaxInt32 {
+			return Value{}, rangeErr(x)
+		}
+		return Value{int32(u)}, nil
+	case int, int8, int16, int64:
+		i := reflect.ValueOf(x).Int()
+		if i > math.MaxInt32 || i < math.MinInt32 {
+			return Value{}, rangeErr(x)
+		}
+		return Value{int32(i)}, nil
+	case float32, float64:
+		f := reflect.ValueOf(x).Float()
+		if f > math.MaxInt32 || f < math.MinInt32 {
+			return Value{}, rangeErr(x)
+		}
+		return Value{int32(f)}, nil
+
+	case URIRef:
+		return Value{x.URL}, nil
+	case URI:
+		return Value{x.URL}, nil
+	case URLRef:
+		return Value{x.URL}, nil
+
+	case Timestamp:
+		return Value{x.Time}, nil
+
+	case *url.URL, *URIRef, *URLRef, *URI, *Timestamp, *time.Time:
+		rx := reflect.ValueOf(x)
+		if rx.IsNil() {
+			return Value{}, valueErr("", x)
+		}
+		return ValueOf(rx.Elem().Interface())
+
+	default:
+		return Value{}, fmt.Errorf("%T is not a CloudEvents type", x)
+	}
+}
+
+func ValueOfBool(x bool) Value           { return Value{x} }
+func ValueOfInteger(x int32) Value       { return Value{x} }
+func ValueOfString(x string) Value       { return Value{x} }
+func ValueOfBinary(x []byte) Value       { return Value{x} }
+func ValueOfURIRef(x URIRef) Value       { return Value{x} }
+func ValueOfTimestamp(x time.Time) Value { return Value{x} }
+
+// String returns the canonical string encoded representation of the value.
+// Use ToString() to extract the value only if it is of type string.
+func (v Value) String() string {
+	switch x := v.Interface().(type) {
+	case bool:
+		return strconv.FormatBool(x)
+	case int32:
+		return strconv.Itoa(int(x))
+	case string:
+		return x
+	case []byte:
+		return base64.StdEncoding.EncodeToString(x)
+	case url.URL:
+		return x.String()
+	case time.Time:
+		return x.UTC().Format(time.RFC3339Nano)
+	case nil:
+		return "<nil>"
```

Comment:
I think this will be an issue for nil things because you might want to omit empty, but it will not be empty with "<nil>"

Reply:
We can make an issue to follow up.

Reply:
I'm not keen on that. The empty string is not a missing value, it is a legitimate value of type string. I agree that some bindings may use/interpret empty strings to mean missing value but that should be handled by code specific to those bindings, not ambiguated in the type system. I'd prefer nil to stringify as a broken, unparseable thing like <nil>. The surrounding code should figure out that it has no value *before* it tries to convert the value it doesn't have :) If it doesn't that's a bug and better fail than cover it up.